plugins/superio/fu-superio-device.c

```diff
@@ -0,0 +1,318 @@
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_superio_device_outb (FuSuperioDevice *self, guint16 port, guint8 data, GError **error)
+{
+	if (pwrite (self->fd, &data, 1, (goffset) port) != 1) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to write to port %04x: %s",
+			     (guint) port,
+			     strerror (errno));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_superio_device_inb (FuSuperioDevice *self, guint16 port, guint8 *data, GError **error)
+{
+	if (pread (self->fd, data, 1, (goffset) port) != 1) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to read from port %04x: %s",
+			     (guint) port,
+			     strerror (errno));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static guint16
+fu_superio_device_probe_id (FuSuperioDevice *self, GError **error)
+{
+	guint8 msb;
+	guint8 lsb;
+
+	if (!fu_superio_device_outb (self, 0x2e, 0x20, error))
+		return FALSE;
+	if (!fu_superio_device_inb (self, 0x2f, &msb, error))
+		return FALSE;
+	if (!fu_superio_device_outb (self, 0x2e, 0x21, error))
+		return FALSE;
+	if (!fu_superio_device_inb (self, 0x2f, &lsb, error))
+		return FALSE;
+	self->id = ((guint16) msb << 8) | (guint16) lsb;
+	return TRUE;
+}
+
+static gboolean
+fu_superio_device_ec_cmd (FuSuperioDevice *self, guint8 data, GError **error)
+{
+	guint8 status = 0x00;
+	g_autoptr(GTimer) timer = g_timer_new ();
+	do {
+		if (!fu_superio_device_inb (self, self->cmd_port, &status, error))
+			return FALSE;
+		if (g_timer_elapsed (timer, NULL) > FU_PLUGIN_SUPERIO_TIMEOUT) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_TIMED_OUT,
+					     "timed out whilst waiting for !busy");
+			return FALSE;
+		}
+	} while ((status & KB_IBF) > 0);
+	return fu_superio_device_outb (self, self->cmd_port, data, error);
+}
+
+static gboolean
+fu_superio_device_ec_read (FuSuperioDevice *self, guint8 *data, GError **error)
+{
+	guint8 status = 0x00;
+	g_autoptr(GTimer) timer = g_timer_new ();
+	do {
+		if (!fu_superio_device_inb (self, self->cmd_port, &status, error))
+			return FALSE;
+		if (g_timer_elapsed (timer, NULL) > FU_PLUGIN_SUPERIO_TIMEOUT) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_TIMED_OUT,
+					     "timed out whilst waiting for !write");
+			return FALSE;
+		}
+	} while ((status & KB_OBF) == 0);
+	return fu_superio_device_inb (self, self->data_port, data, error);
+}
+
+static gboolean
+fu_superio_device_ec_write (FuSuperioDevice *self, guint8 data, GError **error)
+{
+	guint8 status = 0x00;
+	g_autoptr(GTimer) timer = g_timer_new ();
+	do {
+		if (!fu_superio_device_inb (self, self->cmd_port, &status, error))
+			return FALSE;
+		if (g_timer_elapsed (timer, NULL) > FU_PLUGIN_SUPERIO_TIMEOUT) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_TIMED_OUT,
+					     "timed out whilst waiting for !busy");
+			return FALSE;
+		}
+	} while ((status & KB_IBF) > 0);
```

Comment:
Since you're using this twice, maybe this should be it's own function used both by `ec_read` and `ec_write`

Reply:
Done